javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1131 @@
+           with WrappedListIterator[E]{
+      override def set(e: E): Unit = throw new UnsupportedOperationException
+
+      override def add(e: E): Unit = throw new UnsupportedOperationException
+    }
+  }
+
+  private[Collections] object CheckedCollections {
+
+    private final def checkClass[T](elem: T, clazz: Class[T]): Unit = {
+      if (elem != null && !clazz.isInstance(elem))
+        throw new ClassCastException
+    }
+
+    private[Collections] trait CheckedCollection[E, Coll <: Collection[E]]
+        extends WrappedCollection[E, Coll] {
+      protected def elemClazz: Class[E]
+
+      override def add(e: E): Boolean = {
+        checkElem(e)
+        super.add(e)
+      }
+
+      override def addAll(c: Collection[_ <: E]): Boolean = {
+        c.foreach(checkElem)
+        super.addAll(c)
+      }
+
+      protected final def checkElem(elem: E) =
+        CheckedCollections.checkClass(elem, elemClazz)
+    }
+
+    private[Collections] trait CheckedSet[E, Coll <: Set[E]]
+        extends CheckedCollection[E, Coll] with WrappedSet[E, Coll]
+
+    private[Collections] trait CheckedSortedSet[E]
+        extends CheckedSet[E, SortedSet[E]] with WrappedSortedSet[E] {
+
+      override def subSet(fromElement: E, toElement: E): SortedSet[E] =
+        checkedSortedSet(super.subSet(fromElement, toElement), elemClazz)
+
+      override def headSet(toElement: E): SortedSet[E] =
+        checkedSortedSet(super.headSet(toElement), elemClazz)
+
+      override def tailSet(fromElement: E): SortedSet[E] =
+        checkedSortedSet(super.tailSet(fromElement), elemClazz)
+    }
+
+    private[Collections] trait CheckedList[E]
+        extends CheckedCollection[E, List[E]] with WrappedList[E] {
+
+      override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+        c.foreach(checkElem)
+        super.addAll(index, c)
+      }
+
+      override def set(index: Int, element: E): E = {
+        checkElem(element)
+        super.set(index, element)
+      }
+
+      override def add(index: Int, element: E): Unit = {
+        checkElem(element)
+        super.add(index, element)
+      }
+
+      override def listIterator(): ListIterator[E] = listIterator(0)
+
+      override def listIterator(index: Int): ListIterator[E] =
+        new CheckedListIterator[E](inner.listIterator(index), elemClazz)
+
+      override def subList(fromIndex: Int, toIndex: Int): List[E] =
+        checkedList(super.subList(fromIndex, toIndex), elemClazz)
+    }
+
+    private[Collections] trait CheckedMap[K, V, M <: Map[K, V]]
+        extends WrappedMap[K, V, M] {
+      protected def keyClazz: Class[K]
+
+      protected def valueClazz: Class[V]
+
+      override def put(key: K, value: V): V = {
+        checkKeyAndValue(key, value)
+        super.put(key, value)
+      }
+
+      override def putAll(m: Map[_ <: K, _ <: V]): Unit = {
+        m.entrySet().foreach(entry => checkKeyAndValue(entry.getKey, entry.getValue))
```

Comment:
Why not `m.foreach` directly? It seems to me that `entrySet()` is creating an additional indirection for no good reason.